pkg/querier/tripperware/instantquery/instant_query.go

```diff
@@ -68,12 +73,18 @@ func (c instantQueryCodec) DecodeRequest(_ context.Context, r *http.Request, for
 	result.Stats = r.FormValue("stats")
 	result.Path = r.URL.Path
 
-	// Include the specified headers from http request in prometheusRequest.
-	for _, header := range forwardHeaders {
-		for h, hv := range r.Header {
-			if strings.EqualFold(h, header) {
-				result.Headers[h] = hv
-				break
+	isSourceRuler := strings.Contains(r.Header.Get("User-Agent"), tripperware.RulerUserAgent)
+	if isSourceRuler {
+		// When the source is the Ruler, then forward whole headers
+		result.Headers = r.Header
```

Comment:
Nit: Might be better to short circuit and return vs else block

Reply:
What do you mean is like this?
```
if isSourceRuler {
		// When the source is the Ruler, then forward whole headers
		result.Headers = r.Header
		return &result, nil
	} else {
		// Include the specified headers from http request in prometheusRequest.
		for _, header := range forwardHeaders {
			for h, hv := range r.Header {
				if strings.EqualFold(h, header) {
					result.Headers[h] = hv
					break
				}
			}
		}
	}

	return &result, nil
```

Reply:
Something like this

```
if isSourceRuler {
		// When the source is the Ruler, then forward whole headers
		result.Headers = r.Header
		return &result, nil
	} 
// Include the specified headers from http request in prometheusRequest.
for _, header := range forwardHeaders {
	for h, hv := range r.Header {
		if strings.EqualFold(h, header) {
			result.Headers[h] = hv
			break
		}
	}
}
return &result, nil
```